Buttplug.Server/Bluetooth/Devices/MagicMotion.cs

```diff
@@ -4,127 +4,222 @@
-                        $"Index {v.Index} is out of bounds for VibrateCmd for this device.",
-                        Error.ErrorClass.ERROR_DEVICE,
-                        cmdMsg.Id);
-                }
-
-                if (Math.Abs(v.Speed - _vibratorSpeed) < 0.001)
-                {
-                    return new Ok(cmdMsg.Id);
-                }
-
-                _vibratorSpeed = v.Speed;
-            }
-
-            var data = new byte[] { 0x0b, 0xff, 0x04, 0x0a, 0x32, 0x32, 0x00, 0x04, 0x08, 0x00, 0x64, 0x00 };
-            data[9] = Convert.ToByte(_vibratorSpeed * byte.MaxValue);
-
-            // While there are 3 lovense revs right now, all of the characteristic arrays are the same.
-            return await Interface.WriteValue(aMsg.Id,
-                (uint)MagicMotionBluetoothInfo.Chrs.Tx,
-                data, false, aToken);
-        }
-    }
-}
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Buttplug.Core;
+using Buttplug.Core.Messages;
+
+namespace Buttplug.Server.Bluetooth.Devices
+{
+    internal class MagicMotionBluetoothInfo : IBluetoothDeviceInfo
+    {
+        public enum Chrs : uint
+        {
+            Tx = 0,
+        }
+
+        // Rx on other service: 6f468792-f91f-11e3-a847-b2227cce2b54 (6f468bfc-f91f-11e3-a847-b2227cce2b54)
+
+        public Guid[] Services { get; } = { new Guid("78667579-7b48-43db-b8c5-7928a6b0a335") };
+
+        public string[] NamePrefixes { get; } = { };
+
+        public string[] Names { get; } =
+        {
+            "Smart Mini Vibe",
+            "Flamingo",
+            "Eidolon",
+            "Smart Bean", // Kegel Master
+            "Magic Cell", // Dante/Candy
+        };
+
+        public Dictionary<uint, Guid> Characteristics { get; } = new Dictionary<uint, Guid>()
+        {
+            // tx characteristic
+            { (uint)Chrs.Tx, new Guid("78667579-a914-49a4-8333-aa3c0cd8fedc") },
+        };
+
+        public IButtplugDevice CreateDevice(IButtplugLogManager aLogManager,
+            IBluetoothDeviceInterface aInterface)
+        {
+            return new MagicMotion(aLogManager, aInterface, this);
+        }
+    }
+
+    internal class MagicMotion : ButtplugBluetoothDevice
+    {
+        private readonly double[] _vibratorSpeeds = { 0, 0 };
+
+        internal struct MagicMotionType
+        {
+            public string Name;
+            public uint VibeCount;
+            public bool Accelerometer;
+            public bool Pressure;
+            public uint Protocol;
+        }
+
+        internal static readonly Dictionary<string, MagicMotionType> DevInfos =
+            new Dictionary<string, MagicMotionType>()
+            {
+                {
+                    "Smart Mini Vibe",
+                    new MagicMotionType()
+                    {
+                        Name = "Smart Mini Vibe",
+                        VibeCount = 1,
+                        Protocol = 1,

```

Comment:
Protocol should be an enum instead of a number, since that explicitly limits range.